src/main/java/janggi/domain/board/Board.java

```diff
@@ -0,0 +1,85 @@
+package janggi.domain.board;
+
+import janggi.domain.piece.Team;
+import janggi.domain.Turn;
+import janggi.domain.piece.impl.None;
+import janggi.domain.piece.Piece;
+import janggi.domain.piece.Pieces;
+import janggi.domain.piece.Position;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+public class Board {
+    private static final double SCORE_SECOND_TURN_BENEFIT = 1.5;
+
+    private final Map<Position, Piece> board;
+    private final Set<Position> palacePositions;
+
+    public Board(final Map<Position, Piece> board, final Set<Position> palacePositions) {
+        this.board = new HashMap<>(board);
+        this.palacePositions = palacePositions;
+    }
+
+    public Map<Position, Piece> getBoard() {
+        return new HashMap<>(board);
+    }
+
+    public void movePiece(final Team team, final Position beforePosition, final Position afterPosition) {
+        Piece pieceToMove = board.get(beforePosition);
+
+        validateTurn(team, pieceToMove);
+        validateMovement(beforePosition, afterPosition, pieceToMove);
+        executeMovement(beforePosition, afterPosition, pieceToMove);
+    }
+
+    private void validateTurn(final Team team, final Piece piece) {
+        if (!piece.is(team)) {
+            throw new IllegalArgumentException("지금은 " + team.getName() + "팀 기물만 이동할 수 있습니다.");
+        }
+    }
+
+    private void validateMovement(final Position beforePosition, final Position afterPosition, final Piece piece) {
+        if (piece.isPalacePiece() && palacePositions.contains(beforePosition) && palacePositions.contains(
+                afterPosition)) {
+            piece.getPalaceMovableValidator(beforePosition, afterPosition).accept(new Pieces(board));
+            return;
+        }
+        piece.getMovableValidator(beforePosition, afterPosition).accept(new Pieces(board));
```

Comment:
```suggestion
        if (piece.isPalacePiece() && beforePosition.isPalace() && afterPosition.isPalace()) {
            piece.getPalaceMovableValidator(beforePosition, afterPosition).accept(new Pieces(board));
            return;
        }
        piece.getMovableValidator(beforePosition, afterPosition).accept(new Pieces(board));
```

위와 같이 작성되면 어떨까요? 그러면 Board가 아닌 piece에서 궁성용 MovableValidator와 일반용 MovableValidator를 알아서 반환하도록 수정할 수도 있을 것 같습니다.

ex)
```java
void validateMovement(Position beforePosition, Position afterPosition, Piece piece) {
    piece.getMovableValidator(beforePosition, afterPosition).accept(new Pieces(board));
}


// getMovableValidator 구현
public Consumer<Pieces> getMovableValidator(Position beforePosition, Position afterPosition) {
    if (isPalacePiece() && beforePosition.isPalace() && afterPosition.isPalace()) {
        return ... // 궁성용 MovableValidator
    }
    return ... // 일반용 MovableValidator
}
```

Reply:
말씀해주신대로 시도해보려했는데,
`beforePosition.isPalace()`가 가능하려면 Position이 Palace에 대해 알고 있어야하게 됩니다. 
Board -> Position, Position -> Board 이런 식으로 참조하게되는 문제가 발생할 것 같습니다.
palacePositions를 전역적으로 관리해주지 않으면 어떤 식으로든 파라미터가 하나 생겨날 것 같습니다.
다만 가능하다면 Piece에서 입력을 분기처리해주는 것은 좋다고 생각합니다.
`getPalaceMovableValidator`를 부모클래스 Piece에서 정의할 필요가 없어지니까요. 🙂

Reply:
순환 참조 이슈가 있었군요! 저도 몇개 코드만 보고 드는 생각을 전달드린거라 적용 시 다른 이슈가 생길 수 있네요~